mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraIndexTableHandlerTest.java

```diff
@@ -277,6 +303,98 @@ public void updateIndexOnFlagsUpdateShouldDecrementUnseenMessageCountWhenSeenIsS
+                retrieveDeletedMessage(MAILBOX_ID, MessageRange.all())
+                .join()
+                .collect(Guavate.toImmutableList()))
+            .containsExactly(MESSAGE_UID);
+    }
+
+    @Test
+    public void updateIndexOnFlagsUpdateShouldRemoveMessageInDeletedMessageWhenDeletedFlagIsUnset() throws Exception {
+        MailboxMessage message = mock(MailboxMessage.class);
+        when(message.createFlags()).thenReturn(new Flags());
+        when(message.getUid()).thenReturn(MESSAGE_UID);
+        testee.updateIndexOnAdd(message, MAILBOX_ID).join();
+
+        deletedMessageDAO.addDeleted(MAILBOX_ID, MESSAGE_UID).join();
+
+        testee.updateIndexOnFlagsUpdate(MAILBOX_ID, UpdatedFlags.builder()
+            .uid(MESSAGE_UID)
+            .newFlags(new Flags())
+            .oldFlags(new Flags(Flags.Flag.DELETED))
+            .modSeq(MODSEQ)
+            .build()).join();
+
+        assertThat(
+            deletedMessageDAO.
+                retrieveDeletedMessage(MAILBOX_ID, MessageRange.all())
+                .join()
+                .collect(Guavate.toImmutableList()))
+            .isEmpty();
+    }
+
+    @Test
+    public void updateIndexOnFlagsUpdateShouldNotRemoveMessageInDeletedMessageWhenDeletedFlagIsNotUnset() throws Exception {
+        MailboxMessage message = mock(MailboxMessage.class);
+        when(message.createFlags()).thenReturn(new Flags());
+        when(message.getUid()).thenReturn(MESSAGE_UID);
+        testee.updateIndexOnAdd(message, MAILBOX_ID).join();
+
+        deletedMessageDAO.addDeleted(MAILBOX_ID, MESSAGE_UID).join();
+
+        testee.updateIndexOnFlagsUpdate(MAILBOX_ID, UpdatedFlags.builder()
+            .uid(MESSAGE_UID)
+            .newFlags(new Flags())
+            .oldFlags(new Flags(Flags.Flag.SEEN))
+            .modSeq(MODSEQ)
+            .build()).join();
+
+        assertThat(
+            deletedMessageDAO.
+                retrieveDeletedMessage(MAILBOX_ID, MessageRange.all())
+                .join()
+                .collect(Guavate.toImmutableList()))
+            .containsExactly(MESSAGE_UID);
+    }
+
+    @Test
+    public void updateIndexOnFlagsUpdateShouldNotSaveMessageInDeletedMessageWhenDeletedFlagIsNotSet() throws Exception {
+        MailboxMessage message = mock(MailboxMessage.class);
+        when(message.createFlags()).thenReturn(new Flags());
+        when(message.getUid()).thenReturn(MESSAGE_UID);
+        testee.updateIndexOnAdd(message, MAILBOX_ID).join();
+
+        testee.updateIndexOnFlagsUpdate(MAILBOX_ID, UpdatedFlags.builder()
+            .uid(MESSAGE_UID)
+            .newFlags(new Flags(Flags.Flag.RECENT))
+            .oldFlags(new Flags())
+            .modSeq(MODSEQ)
+            .build()).join();
+
+        assertThat(
+            deletedMessageDAO.
+                retrieveDeletedMessage(MAILBOX_ID, MessageRange.all())
+                .join()
+                .collect(Guavate.toImmutableList()))
+            .isEmpty();
+    }
+
```

Comment:
Isn't it what 'updateIndexOnDeleteShouldDeleteMessageFromDeletedMessage' is about ? I guess I am missing something ?